src/Hyperion/Extensions/StreamEx.cs

```diff
@@ -75,15 +74,15 @@ public static void WriteVarint64(this Stream stream, ulong value)
         public static uint ReadUInt16(this Stream self, DeserializerSession session)
         {
             var buffer = session.GetBuffer(2);
-            self.Read(buffer, 0, 2);
+            self.ReadFull(buffer, 0, 2);
```

Comment:
For primitive reads, maybe we should wrap the `Stream` with a `BinaryReader` instead?

Reply:
Reason to not add an additional wrapper is the additional allocation which could hurt the performance.

Reply:
Hmmm... can you make 2 kind of `ReadFull` then? the difference is that one requires that all requested bytes have to be received and it will throw an unexpected EOF exception if it doesn't, the other can read partial request. Lets call them `ReadAll` and `ReadPartial`.
- `ReadAll` must receive all the byes it requested, if we request a read of 8 bytes, then all 8 bytes must be received, it throws if it doesnt. This is used for all primitive reads, such as byte, short, ushort, int, uint, long, etc.
- `ReadPartial` can receive partial result, if we requested 100 bytes, its fine if it received 10. This is used for all unknown length read, such as array and string.

Reply:
You can also do this by adding a flag parameter `allowPartial` to your method, make it false as default. If `allowPartial` is false, then the read methos should throw an EOF exception if the stream returns 0 on read before all bytes are read, and just breaks if `allowPartial` is true

Reply:
all usages of "ReadFull()" requires all requested bytes to be read so the default behavior is throwing an EOF exception for now. If a use case for useful partial reads comes up this flag can be added.